File: 637-average-of-levels-in-binary-tree/637-average-of-levels-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        queue = deque()
        ans = []
        
        queue.append(root)
        
        while queue:
            levelSize = len(queue)
            levelSum = 0
            
            for _ in range(levelSize):
                currNode = queue.popleft()
                
                levelSum += currNode.val
                
                if currNode.left != None:
                    queue.append(currNode.left)
                if currNode.right != None:
                    queue.append(currNode.right)
            ans.append(levelSum / levelSize)
        return ans

File: 637-average-of-levels-in-binary-tree/637-average-of-levels-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        return solve(root);
    }
private:
    vector<double> solve(TreeNode* root) {
        queue<TreeNode*> queue;
        vector<double> ans;
        
        queue.push(root);
        
        while(!queue.empty()) {
            int levelSize = queue.size();
            double levelSum = 0;
            int n = levelSize;
            
            while(levelSize--) {
                auto currNode = queue.front();
                queue.pop();
                
                levelSum += currNode -> val;
                
                if(currNode -> left != nullptr)
                    queue.push(currNode -> left);
                if(currNode -> right != nullptr)
                    queue.push(currNode -> right);
            }
            ans.push_back(levelSum/n);
        }
        return ans;
    }
};